support-lib/cpp/Future.hpp

```diff
@@ -172,8 +172,10 @@ class PromiseBase {
             sharedState->value = std::move(val);
         });
     }
-    template <typename E>
-    void setException(const E& ex) {
+
+    // Lock down to std::exception as a base class, as otherwise make_exception_ptr will happily take other
+    // pointers, and we don't want to encourage rejecting a promise with something other than an exception object.
+    void setException(const std::exception& ex) {
```

Comment:
This will slice the exception object and lose information.  `make_exception_ptr` makes a copy, and since the parameter type is std::exception&, only the std::exception part will be kept.  The template here is just an attempt to retain the full exception.

Reply:
Hmm okay. What annoyed me with this template is I was able to pass in any pointer data. Like a string literal. I want the compiler to stop me. Any ideas how we can prevent that and not lose the derived class information?

Reply:
Sorry that was me, accidentally commented in my personal account. 

We can tighten the parameter requirement by rejecting types that are not `std::is_convertible<std::exception>` with a static_assert.

Reply:
okay I will try that out